ed_sensor_integration/tools/fitter_viz.cpp

```diff
@@ -0,0 +1,382 @@
+        fitter.processSensorData(*snapshot.image, snapshot.sensor_pose, fitterdata);
+
+        bool estimateEntityPose = fitter.estimateEntityPose(fitterdata, snapshot.world_model, entity_id, e->pose(), fitted_pose);
+
+        // poses for visualization
+        geo::Transform2 sensor_pose2d = fitterdata.sensor_pose_xya.projectTo2d();
+        geo::Transform2 entity_pose2d = e->pose().projectTo2d();
+        geo::Transform2 fitted_pose2d = fitted_pose.projectTo2d();
+
+        // show snapshot
+        cv::Mat rgbcanvas = snapshot.image->getRGBImage().clone();
+        cv::imshow("RGB", rgbcanvas);
+
+        // visualize fitting
+        cv::Mat canvas = cv::Mat(canvas_height, canvas_width, CV_8UC3, cv::Scalar(0, 0, 0));
+
+        // paint to screen the location of the sensor
+        cv::circle(canvas, sensorlocation, 3, sensor_colour, cv::FILLED);
+
+        // paint entity (from worldmodel)
+        EntityRepresentation2D entity_2d = fitter.GetOrCreateEntity2D(e);
+        geo::Transform2 relpose = sensor_pose2d.inverse() * entity_pose2d;
+        drawShape2D(canvas, entity_2d.shape_2d, relpose, canvas_resolution, sensor_x, sensor_y, entity_colour);
+
+        if (estimateEntityPose)
+        {
+            // paint fitted entity
+            geo::Transform2 fitted_relpose = sensor_pose2d.inverse() * fitted_pose2d;
+            drawShape2D(canvas, entity_2d.shape_2d, fitted_relpose, canvas_resolution, sensor_x, sensor_y, fitted_colour);
+        }
+        else
+        {
+            std::cout << "Fitted entity not printed in image because of a fitter error" << std::endl;
+        }
+
+        // paint sensor_ranges
+        for (unsigned int i = 0; i < fitterdata.sensor_ranges.size(); ++i)
+        {
+            float fx_ = 2 * fitterdata.sensor_ranges.size() / 4;
+            float x_m = fitterdata.sensor_ranges[i] * ((static_cast<int>(i) - (fitterdata.sensor_ranges.size() / 2)) / fx_);
+            float y_m = fitterdata.sensor_ranges[i];
+
+            // postion to pixels
+            int x_p = sensor_x + (int)(x_m * canvas_resolution);
+            int y_p = sensor_y - (int)(y_m * canvas_resolution);
+
+            if (x_p < 0 || x_p >= canvas_width)
+                continue;
+            if (y_p < 0 || y_p >= canvas_height)
+                continue;
+            if (fitterdata.sensor_ranges[i] == 0) // filter out sensor_ranges equal to zero
+                continue;
+
+            // paint to screen
+            cv::Point centerCircle(x_p, y_p);
+            cv::circle(canvas, centerCircle, 2, measurement_colour, cv::FILLED);
+        }
+
+        cv::imshow("Fitting", canvas);
+        char key = cv::waitKey();
+
+        if (key == 81)  // Left arrow
+        {
+            if (i_snapshot > 0)
+                --i_snapshot;
+        }
+        else if (key == 83) // Right arrow
+        {
+            ++i_snapshot;
+        }
+        else if (key == 'q')
+        {
+            break;
+        }
+    }
```

Comment:
If I understand the code correctly. We are collecting snapshot and showing them in the same loop. By doing this in the same loop, we are creating a big complex mess.

We should split the collection of the snapshot and showing the snapshot. Also I got the feeling stuff could be broken into smaller functions. All should improve usability and readability of the code.

Reply:
That should be addressed by #52